versioned/tiered/rocksdb/src/main/java/com/dremio/nessie/versioned/store/rocksdb/RocksDBStore.java

```diff
@@ -0,0 +1,359 @@
+        valueTypeToColumnFamily.values().forEach(cf -> {
+          try {
+            rocksDB.flush(options, cf);
+          } catch (RocksDBException e) {
+            LOGGER.error("Error flushing column family while closing Nessie RocksDB store.", e);
+          }
+          cf.close();
+        });
+      }
+
+      rocksDB.close();
+      rocksDB = null;
+    }
+  }
+
+  @Override
+  public void load(LoadStep loadstep) throws NotFoundException {
+    for (LoadStep step = loadstep; step != null; step = step.getNext().orElse(null)) {
+      final List<LoadOp<?>> loadOps = step.getOps().collect(Collectors.toList());
+      if (loadOps.isEmpty()) {
+        continue;
+      }
+
+      final List<ColumnFamilyHandle> columnFamilies = new ArrayList<>(loadOps.size());
+      final List<byte[]> keys = new ArrayList<>(loadOps.size());
+      loadOps.forEach(op -> {
+        columnFamilies.add(getColumnFamilyHandle(op.getValueType()));
+        keys.add(op.getId().toBytes());
+      });
+
+      final List<byte[]> reads;
+      try {
+        reads = rocksDB.multiGetAsList(columnFamilies, keys);
+      } catch (RocksDBException e) {
+        throw new RuntimeException(e);
+      }
+
+      if (reads.size() != loadOps.size()) {
+        throw new NotFoundException(String.format("[%d] object(s) missing in load.", loadOps.size() - reads.size()));
+      }
+
+      for (int i = 0; i < reads.size(); ++i) {
+        final LoadOp<?> loadOp = loadOps.get(i);
+        if (null == reads.get(i)) {
+          throw new NotFoundException(String.format("Unable to find requested ref with ID: %s", loadOp.getId()));
+        }
+
+        @SuppressWarnings("rawtypes") final ValueType type = loadOp.getValueType();
+        RocksSerDe.deserializeToConsumer(type, reads.get(i), loadOp.getReceiver());
+        loadOp.done();
+      }
+    }
+  }
+
+  @Override
+  public <C extends BaseValue<C>> boolean putIfAbsent(SaveOp<C> saveOp) {
+    final ColumnFamilyHandle columnFamilyHandle = getColumnFamilyHandle(saveOp.getType());
+    try {
+      final Transaction transaction = transactionDB.beginTransaction(new WriteOptions(), new OptimisticTransactionOptions());
+      // Get exclusive access to key if it exists.
+      final byte[] buffer = transaction.getForUpdate(new ReadOptions(), columnFamilyHandle, saveOp.getId().toBytes(), true);
+      if (null == buffer) {
+        transaction.put(columnFamilyHandle, saveOp.getId().toBytes(), RocksSerDe.serializeWithConsumer(saveOp));
+        transaction.commit();
+        return true;
+      }
+      // Id already exists.
+      return false;
+    } catch (RocksDBException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public <C extends BaseValue<C>> void put(SaveOp<C> saveOp, Optional<ConditionExpression> condition) {
+    if (condition.isPresent()) {
+      RocksBaseValue consumer = RocksSerDe.getConsumer(saveOp);
+      loadSingle(saveOp.getType(), saveOp.getId(), (C) consumer);
```

Comment:
How is this supposed to work in multi-threaded environments? That was a key consideration for this as we discussed before.

Reply:
This will be made transactional.

Reply:
This is the final comment left to implement. Doing that today (Feb 8th)

Reply:
Done.